homeassistant/helpers/device_registry.py

```diff
@@ -283,19 +316,45 @@ def _async_update_device(
     @callback
     def async_remove_device(self, device_id: str) -> None:
         """Remove a device from the device registry."""
-        del self.devices[device_id]
+        device = self.devices.pop(device_id)
+        self.deleted_devices[device_id] = DeletedDeviceEntry(
+            config_entries=set(device.config_entries),
+            connections=list(device.connections),
+            identifiers=list(device.identifiers),
+            id=device.id,
+        )
         self.hass.bus.async_fire(
             EVENT_DEVICE_REGISTRY_UPDATED, {"action": "remove", "device_id": device_id}
         )
         self.async_schedule_save()
 
+    @callback
+    def async_purge_deleted_device(self, device_id: str) -> None:
+        """Permanently remove a device from the device registry."""
+        del self.deleted_devices[device_id]
+        self.async_schedule_save()
+
+    @callback
+    def async_restore_device(self, device_id: str) -> None:
+        """Restore a deleted device."""
+        deleted_device = self.deleted_devices.pop(device_id)
+        self.devices[device_id] = DeviceEntry(
```

Comment:
We should put this logic inside `DeletedDeviceEntry`: `deleted_device.to_device_entry()`

Then this function is small enough that you don't need to break it out into its own function.